### ISSUE TITLE
Out of memory error on MATLAB while running RELAX

### ISSUE THREAD

**Issue by chengq19** [NeilwBailey/RELAX#15]
Hi!!! I'm using the RELAX pipeline to automate preprocesing of ERPs for my data on Ubuntu focal. However, for some reason, parpool keeps shutting down MATLAB when it got to the the stage where MWF is used to clean up EEG artifacts. I checked my kernel log and noticed that the shutdown was due to Out of Memory error and I also noticed that this is likely due to parpool. Hence, I tried to optimise my parallel processing profile (restricting the number of workers) to avoid OOM error. However, my attempts were unsuccesful thus far and the thing is that my desktop specs are pretty good (13th gen i7 processsor with 6 cores i.e., 12 threads, 32 gb RAM etc). I am a bit stuck on how best to resolve the issue and would like to look for some advice on the following:

1) Is there any way to switch off parpool to bypass the OOM error?
2) Is there any way to optimise parpool to avoid OOM error (besides physically increasing the amount of RAM on my computer)?

Thank you!

Best Regards,
Xiaoqin

**Comment by NeilwBailey** [comment]
Dear Xiaoqin

If the error is indeed due to parpool, one solution could be to determine
exactly which function is being implemented when the error occurs, and
replace "parfor" with "for" in that function. This will run the processing
within that loop in serial instead of parallel, avoiding an issue with
parpool. A couple of functions that include parfor are: PREP's
"findNoisyChannels" function, and RELAX's
"RELAX_excluding_channels_and_epoching", "RELAX_excluding_extreme_values"
and "RELAX_muscle" functions.

However, those functions are all implemented before the MWF cleaning step,
and the MWF cleaning step doesn't use any parfor loops. The MWF step does
however sometimes cause out of memory issues if the file size being
processed is very large, or a very long delay period is set. Given your
computer specifications, if you're setting a delay period of more than 16,
then this could likely be a cause of the issue, or if you have very long
files, even a delay period of more than 10 could be causing an issue. In
this case, setting a shorter delay period might help.

Alternatively, if your files are very long (> 10 minutes) with a high
sampling rate (1000Hz), even short delay periods may require too much
memory. You could downsample your data (to ~500Hz) before cleaning (make
sure you lowpass filter first to avoid anti-aliasing effects though - this
choice is available in the latest version RELAX). Or, if your data are
reasonably clean already, you could avoid the MWF cleaning and simply clean
your data with wICA (which produces good outcomes if the data is reasonably
good even before you submit it for cleaning, or if you have a lot of data
for averaging with an ERP or time-frequency analysis type approach).

Hopefully one of these options solves your issue, but feel free to follow
up with me if you have any further queries.

Kind regards,

Neil

**Comment by chengq19** [comment]
Dear Neil,

Thanks so much for your advice and so sorry for only replying when this is marked by you as completed/closed. I tried out a few things and thought to just list down what resolve the issue for me in case it will help other people. So in my case, the issue was because my file size is huge (I had a 1 hour 20 minutes recording acquired at 500 Hz) and the delay period set was too large. So I had to delete data segments that I am not analysing (e.g., EEG data recorded during breaks) and reduced the delay from the default 8 to 6 so the MWF can run. I now use a delay period of 5 and when I manually scanned through the data, it seems that that does already provide a pretty good cleaning. Also, I have to add that because PREP already call parfor, in my case, it was essential to shut down the parallel pool called during PREP findNoisyChannels and have fieldtrip call the parallel pool again so that workers do not get aborted or lose connection during parallel processing when fieldtrip functions are called. Hope this can be useful!

Again, thanks so much for the prompt reply and sorry for my delay in replying!!!

Best Regards,
Xiaoqin
